File: architecture/demo/src/main/java/com/thepacific/clean/MainActivity.java
package com.thepacific.clean;

import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;
import com.thepacific.presentation.common.RouterUtil;
import com.thepacific.presentation.core.Activity;
import com.thepacific.presentation.core.ViewModel;
import javax.inject.Inject;

public class MainActivity extends Activity {

  @Inject
  App app;

  MainViewModel model;

  ProgressBar progressBar;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    model = fetchViewModel();
    progressBar = findViewById(R.id.progressbar);
    findViewById(R.id.button).setOnClickListener(i -> model.load(0).subscribe(it -> {
      switch (it.status) {
        case IN_PROGRESS:
          progressBar.setVisibility(View.VISIBLE);
          break;
        case IRRELEVANT:
          progressBar.setVisibility(View.GONE);
          break;
        case ERROR:
          progressBar.setVisibility(View.GONE);
          break;
        case SUCCESS:
          progressBar.setVisibility(View.GONE);
          RouterUtil.showDialogFragment(getSupportFragmentManager(), BottomSheet.newInstance());
          break;
        default:
          throw new UnsupportedOperationException();
      }
    }));
  }

  @Override
  protected Class<? extends ViewModel> modelClass() {
    return MainViewModel.class;
  }
}

File: architecture/example/src/main/java/com/pacific/example/MainActivity.kt
package com.pacific.arch.example

import android.os.Bundle
import android.widget.Toast
import com.baidu.mapapi.model.LatLng
import com.baidu.mapapi.search.core.SearchResult
import com.baidu.mapapi.search.geocode.*
import com.bumptech.glide.Glide
import com.pacific.arch.presentation.Activity
import com.pacific.arch.presentation.activityViewModel
import com.pacific.example.MainViewModel
import javax.inject.Inject


class MainActivity : Activity() {

    @Inject
    lateinit var app: App

    private val model by activityViewModel(MainViewModel::class.java)

    private val geoCoder: GeoCoder = GeoCoder.newInstance()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        geoCoder.setOnGetGeoCodeResultListener(object : OnGetGeoCoderResultListener {
            override fun onGetGeoCodeResult(result: GeoCodeResult?) {
                if (result == null || result.error == SearchResult.ERRORNO.PERMISSION_UNFINISHED) {
                    toast("没有检索到经纬度,请确保网络正常")
                    return
                }
                if (result.error == SearchResult.ERRORNO.NO_ERROR) {
                    toast("经纬度(${result.location.latitude},${result.location.longitude})")
                    return
                }

                toast("检索出错：${result.error}")
            }

            override fun onGetReverseGeoCodeResult(result: ReverseGeoCodeResult?) {
                if (result == null || result.error == SearchResult.ERRORNO.PERMISSION_UNFINISHED) {
                    toast("没有解析到地址,请确保网络正常")
                    return
                }
                if (result.error == SearchResult.ERRORNO.NO_ERROR) {
                    toast("地址:${result.addressDetail.countryName},"
                            + "${result.addressDetail.province},"
                            + "${result.addressDetail.city}")
                    return
                }

                toast("解析出错：${result.error}")
            }
        })

        geoCoder.geocode(GeoCodeOption().city("北京").address("海淀区上地十街10号"))

        geoCoder.reverseGeoCode(ReverseGeoCodeOption().newVersion(1).location(LatLng(40.00, 118.00)))
    }

    override fun onDestroy() {
        super.onDestroy()
        geoCoder.destroy()
    }

    fun toast(msg: String) {
        Toast.makeText(this, msg, Toast.LENGTH_LONG).show()
    }
}
